pkg/sql: implement levenshtein

CyrusJavan opened the pull request:
Added levenshtein builtin function

Release justification: low-risk update to new functionality
Release note (sql change): Implemented levenshtein(text, text) int.

Resolves: #56820

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56843)
<!-- Reviewable:end -->

jordanlewis reviewed (changes requested):
Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-:-MMRR3m768-EodffH374:b-ucfknh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @CyrusJavan and @jordanlewis)

___
*[pkg/sql/sem/builtins/builtins.go, line 2950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MMRHsX7DYGXyGgNFlIA:-MMRHsX8CMn6qhaJeW80:bjlszf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f14d58dc39dab833187e37d2e9768b5ca5865f0/pkg/sql/sem/builtins/builtins.go#L2950)):*
> ```Go
> 			Types:      tree.ArgTypes{{"source", types.String}, {"target", types.String}},
> 			ReturnType: tree.FixedReturnType(types.Int),
> 			Fn: func(evalCtx *tree.EvalContext, args tree.Datums) (_ tree.Datum, err error) {
> ```

Postgres restricts users from running this on strings that are more than 255 characters long, since you can easily peg a server by running some giant n^2 algorithms. I think we should do exactly the same thing here - return an error: `pgerror.New(pgcode.InvalidParameterValue, "levenshtein argument exceeds maximum length of 255 characters")`
___
*[pkg/util/leven/leven.go, line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MMRFsuq9ceWegUs2e2w:-MMRFsuq9ceWegUs2e2x:b-szk0c2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f14d58dc39dab833187e37d2e9768b5ca5865f0/pkg/util/leven/leven.go#L7)):*
> ```Go
> func Distance(s, t string) int {
> 	lenS, lenT := len(s), len(t)
> 	a, b := make([]int, lenT+1), make([]int, lenT+1)
> ```

Add some comments - what are the arrays doing?
___
*[pkg/util/leven/leven_test.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MMRGAKa1ZIyybgKTox8:-MMRGAKb1DfFts2MMQ19:b-sztk73) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f14d58dc39dab833187e37d2e9768b5ca5865f0/pkg/util/leven/leven_test.go#L28)):*
> ```Go
> 		{
> 			Source:   "123456789",
> 			Target:   "123",
> ```

What about some test cases where the source is suffix of the target, or contained within the target, and vice versa?


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
Thanks for your work on this! We try to be a little careful about which library dependencies we introduce to this project, especially for user-facing functionality like this. It adds a bit of overhead of making sure licensing, correctness, performance, etc all are acceptable, and potentially a burden of needing to update dependencies.

In this case, the algorithm seems simple enough to implement that we could just write the code in cockroachdb. Would you mind going back to the approach you had in your first PR revision? It looked like you were very close with that approach!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-:-MMxlbRX1dZx2hpL1WL-:b-w82pp6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @CyrusJavan and @jordanlewis)



<!-- Sent from Reviewable.io -->

CyrusJavan committed:
pkg/sql: implement levenshtein

Added levenshtein builtin function

Release justification: low-risk update to new functionality
Release note (sql change): Implemented levenshtein(text, text) int.

Resolves: #56820

CyrusJavan commented:
@jordanlewis @rafiss Hello! Hope everyone had a good holiday! I have removed the outside dependency and re-added my own implementation of the Levenshtein distance function (with modifications to properly handle multi-byte characters).

rafiss commented:
this looks great! thanks.

the build is failing because we are also checking to make sure the project can be built with bazel.

can you run `make bazel-generate`?
or manually apply this diff?
```
[23:47:16]
[Step 6/6] --- a/pkg/sql/sem/builtins/BUILD.bazel
[23:47:16]
[Step 6/6] +++ b/pkg/sql/sem/builtins/BUILD.bazel
[23:47:16]
[Step 6/6] @@ -66,6 +66,7 @@ go_library(
[23:47:16]
[Step 6/6]          "//pkg/util/humanizeutil",
[23:47:16]
[Step 6/6]          "//pkg/util/ipaddr",
[23:47:16]
[Step 6/6]          "//pkg/util/json",
[23:47:16]
[Step 6/6] +        "//pkg/util/leven",
[23:47:16]
[Step 6/6]          "//pkg/util/log",
[23:47:16]
[Step 6/6]          "//pkg/util/mon",
[23:47:16]
[Step 6/6]          "//pkg/util/ring",
```

rafiss reviewed (changes requested):
marking as requested changes to address above

rafiss commented on pkg/util/leven/leven.go:
```diff
@@ -0,0 +1,56 @@
+package leven
```
actually, as long as you need to refresh the bazel dependencies, could you move this to a package named `./pkg/util/fuzzystrmatch`? that way other builtins like the one defined in #57615 can be added there too

the function should be renamed to `LevenshteinDistance`

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56843)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @CyrusJavan and @jordanlewis)

___
*[pkg/sql/sem/builtins/builtins.go, line 2935 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MNz8Xjx2GgDGgjfRTRz:-MNz8Xjy0PErh_Spw-Pn:b-anh2e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc076fabe3c98b30826f1ed947edebec2c67c3f/pkg/sql/sem/builtins/builtins.go#L2935)):*
> ```Go
> 	"difference": makeBuiltin(tree.FunctionProperties{UnsupportedWithIssue: 56820, Category: categoryFuzzyStringMatching}),
> 	"levenshtein": makeBuiltin(defProps(),
> 		tree.Overload{
> ```

can you define the overloaded functions as well? according to https://www.postgresql.org/docs/current/fuzzystrmatch.html, PostgreSQL also has

```
levenshtein(text source, text target, int ins_cost, int del_cost, int sub_cost) returns int
levenshtein_less_equal(text source, text target, int ins_cost, int del_cost, int sub_cost, int max_d) returns int
levenshtein_less_equal(text source, text target, int max_d) returns int
```

`makeBuiltin` accepts a variadic parameter for all the different overloads


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56843)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @CyrusJavan and @jordanlewis)

___
*[pkg/sql/logictest/testdata/logic_test/fuzzystrmatch, line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MNzEAHZ6GnUVwpduR99:-MNzEAHZ6GnUVwpduR9A:bqet6po) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc076fabe3c98b30826f1ed947edebec2c67c3f/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L14)):*
> ```
>   (3, '😄', '🐯')
> 
> statement error levenshtein argument exceeds maximum length of 255 characters
> ```

can you add another test for

`select levenshtein(null, 's');` and ` select levenshtein(null, null);


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
the code looks good! there are some lint errors to address, then this can be merged.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-:-MOCmFTm2Flsil6C3hJc:b1l5dv0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @CyrusJavan, @jordanlewis, and @rafiss)

___
*[pkg/sql/sem/builtins/builtins.go, line 2953 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MOCmwPj1PLx5X-bd8KQ:-MOCmwPj1PLx5X-bd8KR:bazq1l1) ([raw file](https://github.com/cockroachdb/cockroach/blob/59223a27f9cdeb25e7ba8b2e8e5aebb163f0331b/pkg/sql/sem/builtins/builtins.go#L2953)):*
> ```Go
> 		tree.Overload{
> 			Types:      tree.ArgTypes{{"source", types.String}, {"target", types.String},
> 				{"ins_cost", types.Int}, {"del_cost", types.Int}, {"sub_cost", types.Int}},
> ```

our format tool is asking for this diff:

```
        -			Types:      tree.ArgTypes{{"source", types.String}, {"target", types.String},
        +			Types: tree.ArgTypes{{"source", types.String}, {"target", types.String},
         				{"ins_cost", types.Int}, {"del_cost", types.Int}, {"sub_cost", types.Int}},
```
___
*[pkg/sql/sem/builtins/builtins.go, line 2967 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MOCn0yWA6k1ArJV-wVm:-MOCn0yWA6k1ArJV-wVn:b-9ivzz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/59223a27f9cdeb25e7ba8b2e8e5aebb163f0331b/pkg/sql/sem/builtins/builtins.go#L2967)):*
> ```Go
> 			},
> 			Info:       "Calculates the Levenshtein distance between two strings. The cost parameters specify how much to " +
> 				"charge for each edit operation. Maximum input length is 255 characters.",
> ```

our format tool is asking for this:
```
        -			Info:       "Calculates the Levenshtein distance between two strings. The cost parameters specify how much to " +
        +			Info: "Calculates the Levenshtein distance between two strings. The cost parameters specify how much to " +
         				"charge for each edit operation. Maximum input length is 255 characters.",
```
___
*[pkg/util/fuzzystrmatch/leven.go, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MOCmSIv4Y45m2-vPS8b:-MOCmSIv4Y45m2-vPS8c:ba5ymlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/59223a27f9cdeb25e7ba8b2e8e5aebb163f0331b/pkg/util/fuzzystrmatch/leven.go#L1)):*
> ```Go
> package fuzzystrmatch
> ```

the file needs copyright headers. you can just copy it from a different file
___
*[pkg/util/fuzzystrmatch/leven.go, line 10 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MOCmcphCK9nZYT2mkYC:-MOCmcpi7WCuYVH-FIlm:b-z43imo) ([raw file](https://github.com/cockroachdb/cockroach/blob/59223a27f9cdeb25e7ba8b2e8e5aebb163f0331b/pkg/util/fuzzystrmatch/leven.go#L10)):*
> ```Go
> // with O(n^2) time complexity and O(n) space complexity.
> func LevenshteinDistanceWithCost(source, target string,
> 	insCost, delCost, subCost int) int {
> ```

our format tool doesn't like the indentation here. i think the function declaration can all be in one line

to automate this, run 
`        bin/crlfmt "-fast" "-ignore" "\\.(pb(\\.gw)?)|(\\.[eo]g)\\.go|/testdata/|^sql/parser/sql\\.go$|_generated\\.go$" "-tab" "2" "-w" "/go/src/github.com/cockroachdb/cockroach/pkg"`
___
*[pkg/util/fuzzystrmatch/leven_test.go, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-MOCmT1L4vZH5tvRlyeD:-MOCmT1L4vZH5tvRlyeE:ba5ymlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/59223a27f9cdeb25e7ba8b2e8e5aebb163f0331b/pkg/util/fuzzystrmatch/leven_test.go#L1)):*
> ```Go
> package fuzzystrmatch
> ```

the file needs copyright headers. you can just copy it from a different file


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thanks so much for your contribution!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56843#-:-MOEsW2E0uP3JNWIfOQq:b-4iftxt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @CyrusJavan and @jordanlewis)



<!-- Sent from Reviewable.io -->

rafiss commented:
bors r+
